src/django_idom/hooks.py

```diff
@@ -30,3 +54,117 @@ def use_websocket() -> IdomWebsocket:
+
+
+def use_query(
+    query: Callable[_Params, _Result | None],
+    *args: _Params.args,
+    **kwargs: _Params.kwargs,
+) -> Query[_Result | None]:
+    query_ref = use_ref(query)
+    if query_ref.current is not query:
+        raise ValueError(f"Query function changed from {query_ref.current} to {query}.")
+
+    should_execute, set_should_execute = use_state(True)
+    data, set_data = use_state(cast(Union[_Result, None], None))
+    loading, set_loading = use_state(True)
+    error, set_error = use_state(cast(Union[Exception, None], None))
+
+    @use_callback
+    def refetch() -> None:
+        set_should_execute(True)
+        set_data(None)
+        set_loading(True)
+        set_error(None)
+
+    @use_effect(dependencies=[])
+    def add_refetch_callback() -> Callable[[], None]:
+        # By tracking callbacks globally, any usage of the query function will be re-run
+        # if the user has told a mutation to refetch it.
+        _REFETCH_CALLBACKS[query].add(refetch)
+        return lambda: _REFETCH_CALLBACKS[query].remove(refetch)
+
+    @use_effect(dependencies=None)
+    def execute_query() -> None:
+        if not should_execute:
+            return
+
+        def thread_target() -> None:
+            try:
+                query_result = query(*args, **kwargs)
+            except Exception as e:
+                set_data(None)
+                set_loading(False)
+                set_error(e)
+                return
+            finally:
+                set_should_execute(False)
+
+            set_data(query_result)
+            set_loading(False)
+            set_error(None)
+
+        # We need to run this in a thread so we don't prevent rendering when loading.
+        # We also can't do this async since Django's ORM doesn't support this yet.
+        Thread(target=thread_target, daemon=True).start()
+
+    return Query(data, loading, error, refetch)
+
+
+def use_mutation(
+    mutate: Callable[_Params, None],
+    refetch: Callable[..., Any] | Sequence[Callable[..., Any]],
+) -> Mutation[_Params]:
+    loading, set_loading = use_state(True)
+    error, set_error = use_state(cast(Union[Exception, None], None))
+
+    @use_callback
+    def call(*args: _Params.args, **kwargs: _Params.kwargs) -> None:
+        set_loading(True)
+
+        def thread_target() -> None:
+            try:
+                mutate(*args, **kwargs)
+            except Exception as e:
+                set_loading(False)
+                set_error(e)
+            else:
+                set_loading(False)
+                set_error(None)
+                for query in (refetch,) if callable(refetch) else refetch:
+                    for callback in _REFETCH_CALLBACKS.get(query) or ():
+                        callback()
+
+        # We need to run this in a thread so we don't prevent rendering when loading.
+        # We also can't do this async since Django's ORM doesn't support this yet.
+        Thread(target=thread_target, daemon=True).start()
```

Comment:
I believe we may as well do this async, using [`database_sync_to_async`](https://channels.readthedocs.io/en/stable/topics/databases.html).

This would also cover us in automatically closing out the connection to prevent other threads from reusing it.